Bluetooth: Controller: Fix overlapping auxiliary advertising events

cvinayak opened the pull request:
When multiple advertising sets are support then use advanced
scheduling implementation to place multiple auxiliary PDUs
from multiple advertising sets in a non-overlapping events.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix overlapping advertising events

When multiple advertising sets are support then use advanced
scheduling implementation to place multiple auxiliary PDUs
and periodic advertising PDUs in a non-overlapping schedule.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_pdu.h:
```diff
@@ -101,7 +101,8 @@ static inline void lll_adv_aux_data_enqueue(struct lll_adv_aux *lll,
 }
 
-static inline struct pdu_adv *lll_adv_aux_data_peek(struct lll_adv_aux *lll)
+static inline struct pdu_adv *
+	lll_adv_aux_data_peek(const struct lll_adv_aux *const lll)
 {
 	return (void *)lll->data.pdu[lll->data.last];
 }
```
Isn't using `const` here misleading, as we are returning a non-const pointer from the struct?

Thalley replied:
I am not sure I followed that at all. 

The issue here is that we expose the function with a `const` parameter, which typically means that the caller can be sure that the parameter's values are not changed (of course, with a cast to `(void *)` we can always do what we want in C). 

Since we are returning a field from the struct, as non-const, the `const` argument qualifier not only has no meaning (as we can actually change a field in the `const` parameter afterwards), but is misleading, and serves no purpose.

ppryga replied:
@Thalley is right. You just `un-const` part of a structure inside the function and return it as if it were never const. That can break something in a nasty way. This is actually breaking a chain objects that are `const` or are pointed by pointers to `const`.

Thalley replied:
Allow me to present an sample example of why I think this is an issue:

```c
#include <stdio.h>

struct my_struct {
	int a;
};

static int *foo(const struct my_struct *const ms)
{
	return (void *)&ms->a;
}

static void bar(const struct my_struct *const ms)
{
	int *a = foo(ms);

	/* We can now modify a */
	*a = 15;
}

int main()
{
	struct my_struct ms = { .a = 10 };

	printf("a: %d\n", ms.a); /* prints 10 */

	/* Calling bar expecting no change since argument is const */
	bar(&ms);

	printf("a: %d\n", ms.a); /* prints 15 */

	return 0;
}
```

If you need to cast it to a non-const pointer in other to build it, then it sounds like the callers of this function can also not use `const` as the content of the `pdu` (a field of the `struct lll_adv_aux`) can be/is changed.

cvinayak replied:
What `lll->data.pdu[lll->data.last]` points to is never a `const`, it is a pointer that is allocated with PDU buffers from a pool and PDU contents are always modified. The pointer stored is qualified as constant by the change in this PR, but the contents are not constant.

https://github.com/zephyrproject-rtos/zephyr/blob/55be9786bcf0c5b3ba4e04fe8f4df8f0f4e8febc/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_types.h#L15-L18

Thalley replied:
You are absolutely correct Vinayak :) Please excuse my lack of knowledge of the LLL data structures (I expected the `pdu_adv` to be part of the `lll_adv_aux`, and not just be a reference kept in the `data` array). 

This should be fine

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -743,6 +793,32 @@ static void after_cen_offset_get(uint16_t conn_interval, uint32_t ticks_slot,
+	ARG_UNUSED(ticks_to_expire);
+	ARG_UNUSED(op_context);
+
+	return (ticker_id >= TICKER_ID_ADV_AUX_BASE) &&
+	       (ticker_id <= TICKER_ID_ADV_AUX_LAST);
```
Consider `IN_RANGE`:
```suggestion
	return IN_RANGE(ticker_id, TICKER_ID_ADV_AUX_BASE, TICKER_ID_ADV_AUX_LAST);
```

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -1079,23 +1080,34 @@ uint32_t ull_adv_aux_start(struct ll_adv_aux_set *aux, uint32_t ticks_anchor,
+
+	err = ull_sched_after_adv_aux_slot_get(TICKER_USER_ID_THREAD,
+					       ticks_slot, &ticks_anchor_aux);
+	if (!err) {
```
What if there is an error? This it what I would ask if I didn't see the code of the `ull_sched_after_adv_aux_slot_get`. 
The code should not force you to read body every called function to understand what is going on.

I suggest to change the name of the function or do not use `err` as a variable where returned value is assigned if the returned value does not mean an error code.

cvinayak replied:
I do not think the caller functions needs to be read, clearly if there is an error then the return parameter is not used. If there is no free slot (or no existing auxiliary sets), the callee uses the calculations made using current tick is used as anchor reference.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -111,6 +111,42 @@ int ull_sched_after_adv_sync_slot_get(uint8_t user_id, uint32_t ticks_slot_abs,
+	uint32_t ticks_slot;
+	uint8_t ticker_id;
+
+	ticker_id = after_match_slot_get(user_id, ticks_slot_abs,
```
Yes, this function _returns a ticker, its ticks to expire and its ticks slot after it is the matching ticker identified by a function callback_ that is supplied.

I will rename to `match_slot_get` and return `ticker_id`

cvinayak reviewed (commented):
Converting to draft as there is more updates as connections, auxiliary events and periodic events needs to co-operate together in a non-overlapping schedule.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -571,15 +644,19 @@ static uint8_t after_match_slot_get(uint8_t user_id, uint32_t ticks_slot_abs,
+	ticks_anchor_prev = 0U;
 	ticks_to_expire = ticks_to_expire_prev = 0U;
 	ticks_slot_abs_prev = 0U;
+	retry = 4U;
```
I think the code would benefit from having a comment w.r.t. why `4` was chosen for `retry`.

cvinayak replied:
done

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -685,86 +778,71 @@ static void ticker_op_cb(uint32_t status, void *param)
 
-#if defined(CONFIG_BT_CONN)
-static bool ticker_conn_match_op_cb(uint8_t ticker_id, uint32_t ticks_slot,
-				    uint32_t ticks_to_expire, void *op_context)
+static bool ticker_match_op_cb(uint8_t ticker_id, uint32_t ticks_slot,
+			       uint32_t ticks_to_expire, void *op_context)
 {
 	ARG_UNUSED(ticks_slot);
 	ARG_UNUSED(ticks_to_expire);
 	ARG_UNUSED(op_context);
 
-	return (ticker_id >= TICKER_ID_CONN_BASE) &&
-	       (ticker_id <= TICKER_ID_CONN_LAST);
-}
+	return false ||
 
-static struct ull_hdr *conn_ull_hdr_get_cb(uint8_t ticker_id)
-{
-	struct ll_conn *conn;
+#if defined(CONFIG_BT_CTLR_ADV_EXT) && defined(CONFIG_BT_BROADCASTER)
+	       IN_RANGE(ticker_id, TICKER_ID_ADV_AUX_BASE,
+			TICKER_ID_ADV_AUX_LAST) ||
 
-	conn = ll_conn_get(ticker_id - TICKER_ID_CONN_BASE);
-	if (!conn || conn->lll.role) {
-		return NULL;
-	}
+#if defined(CONFIG_BT_CTLR_ADV_PERIODIC)
+	       IN_RANGE(ticker_id, TICKER_ID_ADV_SYNC_BASE,
+			TICKER_ID_ADV_SYNC_LAST) ||
+#endif /* CONFIG_BT_CTLR_ADV_PERIODIC */
+#endif /* CONFIG_BT_CTLR_ADV_EXT && CONFIG_BT_BROADCASTER */
+
+#if defined(CONFIG_BT_CONN)
+	       IN_RANGE(ticker_id, TICKER_ID_CONN_BASE,
+			TICKER_ID_CONN_LAST) ||
+#endif /* CONFIG_BT_CONN */
 
-	return &conn->ull;
+	       false;
```
Do we really need two `false` in this? To me it seems we can remove one of them.

cvinayak replied:
When all the conditional compiles are off, then to match the `return false ||`, the `false;` is needed

Thalley replied:
> When all the conditional compiles are off, then to match the `return false ||`, the `false;` is needed

In the case of all conditional compiles are off, then it's `false || false`, right? 

If we move the `||` to the start of the each conditional, then we can reduce that to simply `false`. 

Either way, it won't matter in the end :)

cvinayak replied:
if `||` is moved to beginning, then `return false` will need a `;` in the end, which appears odd on its own.

cvinayak committed:
Bluetooth: Controller: Use maximum Adv PDU size time reservation

Use the maximum advertising PDU size time reservation
consider the Advertising Data could be updated from zero to
maximum support size. If maximum time reservation is
disabled then time reservation corresponding to the
Advertising Data present at the time of the start/enable
of Advertising is used.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -341,8 +352,8 @@ config BT_CTLR_XTAL_THRESHOLD
 	bool "Advanced scheduling"
-	depends on (BT_MAX_CONN != 0) && BT_CTLR_SCHED_ADVANCED_SUPPORT
-	default y if !(BT_PERIPHERAL && !BT_CENTRAL)
+	depends on BT_CTLR_SCHED_ADVANCED_SUPPORT && (BT_CONN || (BT_CTLR_ADV_EXT && (BT_CTLR_ADV_AUX_SET != 0) && (BT_CTLR_ADV_AUX_SET != 1)) || BT_CTLR_ADV_ISO)
```
Should we worry about the maximum line length of 100 here?

Thalley replied:
It can be done with `\`, e.g.

```
config NRFX_NVMC
	bool "NVMC driver"
	depends on $(dt_has_compat,$(DT_COMPAT_NORDIC_NRF51_FLASH_CONTROLLER)) \
		|| $(dt_has_compat,$(DT_COMPAT_NORDIC_NRF52_FLASH_CONTROLLER)) \
		|| $(dt_has_compat,$(DT_COMPAT_NORDIC_NRF53_FLASH_CONTROLLER)) \
		|| $(dt_has_compat,$(DT_COMPAT_NORDIC_NRF91_FLASH_CONTROLLER))
```

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -60,56 +58,92 @@ static void win_offset_calc(struct ll_conn *conn_curr, uint8_t is_select,
+		if (false) {
 
-		time_us = ull_adv_sync_time_get(sync, PDU_AC_PAYLOAD_SIZE_MAX);
+		} else if (IN_RANGE(ticker_id, TICKER_ID_ADV_AUX_BASE,
+			     TICKER_ID_ADV_AUX_LAST)) {
```
```suggestion
		} else if (IN_RANGE(ticker_id, TICKER_ID_ADV_AUX_BASE,
				    TICKER_ID_ADV_AUX_LAST)) {
```
indent

cvinayak replied:
done

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -60,56 +58,92 @@ static void win_offset_calc(struct ll_conn *conn_curr, uint8_t is_select,
+			return 0;
+
+#if defined(CONFIG_BT_CTLR_ADV_PERIODIC)
+		} else if (IN_RANGE(ticker_id, TICKER_ID_ADV_SYNC_BASE,
+			     TICKER_ID_ADV_SYNC_LAST)) {
```
```suggestion
		} else if (IN_RANGE(ticker_id, TICKER_ID_ADV_SYNC_BASE,
				    TICKER_ID_ADV_SYNC_LAST)) {
```
indent

cvinayak replied:
done

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -560,6 +594,39 @@ static void win_offset_calc(struct ll_conn *conn_curr, uint8_t is_select,
+					      ticks_anchor));
+	}
+
+	if ((*win_offset_us & BIT(31)) == 0) {
```
What is bit 31 here?

cvinayak replied:
Used to test Most Significant Bit of uint32_t. I could not find a suitable define.

Thalley replied:
The comment helps :) So the MSBit indicates that the offset is in the past? 

Sounds like something we could define a (named) check for, but if it is just used here then the comment is good enough.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -571,19 +638,40 @@ static uint8_t after_match_slot_get(uint8_t user_id, uint32_t ticks_slot_abs,
+	 * restarted with the new reference ticks_anchor value.
+	 * Simultaneous continuous scanning on 1M and Coded PHY, alongwith
+	 * directed advertising and one other state/role could expire in quick
+	 * succession, hence have a retry count of 4.
```
I am still not sure why we are using a retry count of 4, compared to e.g. 3 or 5 or some other number. This comment only states that we need a retry, but not really how many.

cvinayak replied:
As mentioned in the comment, retry 4 times due to ticker expiry with 1M, Coded PHY, directed advertising and one other state/role (can be central/peripheral/advertiser/etc.). Preemption of the iterations (while loop) by ISR from these ticker expiry need the loop to restart to find the ticker offsets all over.

Thalley replied:
Ah, that makes sense. Was not aware of that detail :)

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -685,86 +789,96 @@ static void ticker_op_cb(uint32_t status, void *param)
+			TICKER_ID_ADV_AUX_LAST) ||
 
-	conn = ll_conn_get(ticker_id - TICKER_ID_CONN_BASE);
-	if (!conn || conn->lll.role) {
-		return NULL;
-	}
+#if defined(CONFIG_BT_CTLR_ADV_PERIODIC)
+	       IN_RANGE(ticker_id, TICKER_ID_ADV_SYNC_BASE,
+			TICKER_ID_ADV_SYNC_LAST) ||
+#endif /* CONFIG_BT_CTLR_ADV_PERIODIC */
+#endif /* CONFIG_BT_CTLR_ADV_EXT && CONFIG_BT_BROADCASTER */
 
-	return &conn->ull;
+#if defined(CONFIG_BT_CONN)
+	       IN_RANGE(ticker_id, TICKER_ID_CONN_BASE,
+			TICKER_ID_CONN_LAST) ||
+#endif /* CONFIG_BT_CONN */
```
Looks like we can replace the `#if defined` with `IS_ENABLED` here, right?

cvinayak replied:
`TICKER_ID_CONN_BASE` and `TICKER_ID_CONN_LAST` are not defined unless `CONFIG_BT_CONN`

Thalley replied:
Ah, okay, was not aware that we guarded macros by Kconfigs :)

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -685,86 +789,96 @@ static void ticker_op_cb(uint32_t status, void *param)
-	if (err) {
-		return;
-	}
+#if defined(CONFIG_BT_CTLR_ADV_EXT) && defined(CONFIG_BT_BROADCASTER)
```
Can't we use `IS_ENABLED` here?

cvinayak replied:
`TICKER_ID_ADV_AUX_BASE` and `TICKER_ID_ADV_AUX_LAST` are not defined unless `CONFIG_BT_CTLR_ADV_EXT`

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sched.c:
```diff
@@ -685,86 +789,96 @@ static void ticker_op_cb(uint32_t status, void *param)
-	}
+		aux = ull_adv_aux_get(ticker_id - TICKER_ID_ADV_AUX_BASE);
+		if (aux) {
+#if defined(CONFIG_BT_CTLR_ADV_RESERVE_MAX)
```
Can't we use `IS_ENABLED` here?

cvinayak replied:
done

Thalley reviewed (changes requested):
Generally looks good. Some few minor code style comments, and then it looks like quite a few of the `if defined` can be replaced with `IS_ENABLED` which possible can increase readability
